### PR TITLE
UI refactor

### PR DESCRIPTION
**Pull request by MiniaczQ**
Fixes: #30

### CODE REVIEW

File: src/util/ui/widgets.rs [hunk 1]
```diff
@@ -28,58 +47,76 @@ impl<'w, 's> MyWidgets<'w> for Commands<'w, 's> {
...
+/// Widgets spawning trait.
+pub(crate) trait Widgets<'a> {
+    /// Spawns a simple button node with text.
+    fn my_button<I: Into<String>>(&mut self, text: I) -> EntityCommands;
 
+    //// Spawns a simple label.
+    fn my_label<I: Into<String>>(&mut self, text: I) -> EntityCommands;
+}
+
+impl<'a, T: Spawner<'a>> Widgets<'a> for T {
```
Comment (janhohenheim) [inline src/util/ui/widgets.rs:65]:
```suggestion
impl<'a, T> Widgets<'a> for T 
	where T: Spawner<'a>
{
```
Personal preference, feel free to disregard :)

**Review by janhohenheim — CHANGES_REQUESTED** [review]
I'd strongly prefer the `my_` prefix to be gone

File: src/util/ui/widgets.rs (hunk 1 above)
Comment (MiniaczQ) [reply src/util/ui/widgets.rs:65]: removed for widgets, my_root -> ui_root so it's a bit clearer what it is

File: src/util/ui/widgets.rs [hunk 2]
```diff
@@ -1,22 +1,41 @@
...
-    fn my_label<I: Into<String>>(&mut self, text: I) -> EntityCommands;
+impl<'a> Spawner<'a> for ChildBuilder<'a> {
+    fn spawn<B: Bundle>(&mut self, bundle: B) -> EntityCommands {
+        self.spawn(bundle)
+    }
+}
+
+/// Root container spawning trait.
+pub(crate) trait RootContainers {
+    /// Spawns root node that covers full screen and verticaly and horizontaly centers it's content.
```
Comment (janhohenheim) [inline src/util/ui/widgets.rs:32]:
```suggestion
    /// Spawns a root node that covers the full screen
    /// and centers its content horizontally and vertically.
```

**Review by janhohenheim — CHANGES_REQUESTED** [review]

File: src/util/ui/widgets.rs (hunk 1 above)
Comment (janhohenheim) [reply src/util/ui/widgets.rs:64]: Hmm? I believe my suggestion still applies. It's just about the amount of noise that is happening on the line before we get to the name `Widgets`.

File: src/screen/credits.rs [hunk 3]
```diff
@@ -20,15 +20,15 @@ enum CreditsAction {
...
-        .insert(CreditsAction::Back)
-        .set_parent(list);
+        .ui_root()
+        .with_children(|children| {
+            children.label("Alice - Foo");
+            children.label("Bob - Bar");
+
+            children.button("Back").insert(CreditsAction::Back);
+        })
+        .insert(StateScoped(Screen::Credits));
```
Comment (MiniaczQ) [inline src/screen/credits.rs:31]: image

File: src/screen/title.rs [hunk 4]
```diff
@@ -18,16 +18,13 @@ enum TitleAction {
...
     commands
-        .my_button("Credits")
-        .insert(TitleAction::Credits)
-        .set_parent(list);
+        .ui_root()
+        .with_children(|children| {
+            children.button("Play").insert(TitleAction::Play);
+            children.button("Credits").insert(TitleAction::Credits);
+        })
+        .insert(StateScoped(Screen::Title));
```
Comment (janhohenheim) [inline src/screen/title.rs]:
```suggestion
        .ui_root()
        .insert(StateScoped(Screen::Title))
        .with_children(|children| {
            children.button("Play").insert(TitleAction::Play);
            children.button("Credits").insert(TitleAction::Credits);
        });
```

**Review by janhohenheim — CHANGES_REQUESTED** [review]

File: src/util/ui/widgets.rs [hunk 5]
```diff
@@ -30,56 +50,72 @@ impl<'w, 's> MyWidgets<'w> for Commands<'w, 's> {
...
-    fn my_button<I: Into<String>>(&mut self, text: I) -> EntityCommands {
-        let button = self
-            .spawn((ButtonBundle {
+    /// Spawns a simple label.
+    fn label<I: Into<String>>(&mut self, text: I) -> EntityCommands;
+}
+
+impl<'a, T: Spawner<'a>> Widgets<'a> for T {
+    fn button<I: Into<String>>(&mut self, text: I) -> EntityCommands {
+        let mut entity_commands = self.spawn((
```
Comment (benfrankel) [inline src/util/ui/widgets.rs:66]: This could be a single method chain with `self.spawn(...).with_children(...)`. Same for `fn label`.

**Review by janhohenheim — APPROVED** [review]

**Review by benfrankel — CHANGES_REQUESTED** [review]
We'll probably want a `ui/prelude.rs` to re-export extension traits. Doesn't have to be in this PR.

File: src/util/ui/widgets.rs (hunk 5 above)
Comment (MiniaczQ) [reply src/util/ui/widgets.rs:66]: no it cannot, with_children returns mutable reference not ownership, i tried :(

Comment (benfrankel) [reply src/util/ui/widgets.rs:66]: Unfortunate.

**Review by benfrankel — APPROVED** [review]

**Pull request merged**